### PR TITLE
server: fix ~InstanceImpl() segfault

### PR DESCRIPTION
**Pull request by ambuc**
*Description*: This fixes a rare but repeatedly observed segfault, described in detail in the comments of `InstanceImpl::~InstanceImpl()`. 

TL;DR; the `InstanceImpl`'s listener manager instance can potentially contain a target with a callback which, at destruction time (see `RdsRouteConfigSubscription::~RdsRouteConfigSubscription`) goes to unregister it from the top-level `InitManager`, which has already been freed.

In practice, I see this segfault a few times a week, and only ever at teardown (inside `~MainCommonBase`), which could be why it hasn't been crashing anyone.

After a medium amount of trying, I couldn't figure out how to reproduce this in a unit or integration test. A unit test for `//test/server:server_test` would need to stand up a real or mocked `ListenerManager` which owns some target with a callback to `server.initManager()`, which seems very complex (unless I'm overlooking something and there's in-repo precedent for this sort of thing).

*Risk Level*: Medium -- this passes all tests, but what fixes one uncaught bug might cause another.
*Testing*: Al tests pass, no tests changed.
*Docs Changes*: N/A
*Release Notes*: N/A

### CODE REVIEW

**Comment by ggreenway** [comment]
It seems like an integration test that does a shutdown like this may catch this in the ASAN/TSAN builds.

**Comment by htuch** [comment]
I'd like @lizan to also take a look before we consider this finalized, he has the deep knowledge on RDS shutdown from recent PRs and reviews.

**Comment by htuch** [comment]
Otherwise LGTM, thanks for the detective work. I'm surprised we can't create a unit (rather than full system integration) test to mock this up reasonably though.

File: source/server/server.cc [hunk 1]
```diff
@@ -95,6 +95,17 @@ InstanceImpl::~InstanceImpl() {
   // Stop logging to file before all the AccessLogManager and its dependencies are
   // destructed to avoid crashing at shutdown.
   file_logger_.reset();
+
+  // Destruct the ListenerManager explicitly, before InstanceImpl's local init_manager_ is
+  // destructed.
+  //
+  // The ListenerManager's DestinationPortsMap contains FilterChainSharedPtrs. There is a rare race
```
Comment (lizan) [inline source/server/server.cc]: I'm not sure how this is a RARE race, can you elaborate a bit more? Particularly, when RdsRouteConfigSubscription get registered to top-level InitManager?
- FilterChains contains HttpConnectionManager is pretty common
- HttpConnectionManager contains RdsRouteConfigProvider is pretty common too
- Any dynamic RdsRouteConfigProvider contains a RdsRouteConfigSubscription

File: source/server/server.cc [hunk 2]
```diff
@@ -95,6 +95,17 @@ InstanceImpl::~InstanceImpl() {
...
+  // condition where one of these FilterChains contains an HttpConnectionManager, which contains an
+  // RdsRouteConfigProvider, which contains an RdsRouteConfigSubscriptionSharedPtr. Since
+  // RdsRouteConfigSubscription is an Init::Target, ~RdsRouteConfigSubscription triggers a callback
+  // set at initialization, which goes to unregister it from the top-level InitManager, which has
+  // already been destructed (use-after-free) causing a segfault.
+  listener_manager_.reset();
```
Comment (lizan) [inline source/server/server.cc]: While I think this did fix for the case described above, can we move `init_manager_` before `listener_manager` or potentially anything may contain it, in member variable list? InitManagerImpl doesn't seems to have anything depends on others, so it could moved before other stuff.

Comment (ambuc) [reply source/server/server.cc]: I'm happy to move the private member variables around, but without a test it's easy to imagine this accidentally getting undone in the future. A manual `.reset()` seems like a better solution for that fear.

File: source/server/server.cc (hunk 1 above)
Comment (ambuc) [reply source/server/server.cc]: I think the circumstances are common but the resulting segfault is rare. 

I'm not 100% on how memory works here under the hood, but my personal theory is that when the subscription goes to unregister itself from the list of `targets_` inside `InitManagerImpl`, `targets_` has just been freed but not not overwritten (?) so this operation often succeeds. @mrice32 might have more color here.

Comment (lizan) [reply source/server/server.cc]: Yes use-after-free is always rare without ASAN, does that mean you can have a config which always trigger use-after-free with ASAN?

Comment (ambuc) [reply source/server/server.cc]: That's an interesting thought, the answer is surely yes. I'll do some digging and report back.

Comment (lizan) [reply source/server/server.cc]: Once you have that, we could have some sort of test here. Could be an integration test with shutdown or sth similar.

**Comment by lizan** [comment]
Was looking around a way for test, but it was hard to reproduce with minimal config, filed #4302 to track tests. Lets merge this to resolve the real issue first.
